### ISSUE TITLE
Revert #812

### ISSUE THREAD

**Issue by dmwmt** [mobxjs/mobx-react#820]
I believe we need to revert this change: #812 
1. if `process` is not available in the global scope (and it is not in browsers) it will break because you can't check for variable existence by doing it like
```
if (nonExistentVar) { ... }
```
2. It breaks dead-code elimination for some reason, probably because of a more complex expression.

**Comment by FredyC** [comment]
And do you have a suggestion for the proper solution?

**Comment by dmwmt** [comment]
There should be two versions in the package (production and development) built with the corresponding `NODE_ENV` values so these conditions should not be there.
And we should have a bootstrap script, which will choose one of these files. Just like the `react` package does.

**Comment by FredyC** [comment]
Once we migrate the package to TypeScript (#818) we might use a different build tool that handles that more gracefully.

For now, it's probably best if you downgrade to 6.1.4.

**Comment by dmwmt** [comment]
Right now the package is completely broken, as bundlers have nothing to do with the check:
```
if (process ...) { .... }
```
and just because browsers don't have `process` defined, 6.1.5 can't be used in a browser.

**Comment by FredyC** [comment]
I have changed `latest` dist-tag in NPM to 6.1.4, so it won't affect others.

**Comment by dmwmt** [comment]
Tools like `yarn upgrade-interactive` or `npm-upgrade` still suggest 6.1.5 as an upgrade option.

**Comment by FredyC** [comment]
Sorry, this will have to suffice for now. I don't want to poke into code now with such major overwrite and cause conflicts. You can help expedite that eg with reviewing that PR if you know TypeScript.

**Comment by FredyC** [comment]
Btw, I wonder why 6.1.4 has worked with a direct `if (process.env.NODE_ENV !== "production") {`. Why browser doesn't have a problem with that? Are you using UMD build?

**Comment by dmwmt** [comment]
Old version:
```
if (process.env.NODE_ENV !== "production") {
```
`process.env.NODE_ENV` is replaced with the actual string value by the bundler.
Having `process.env.NODE_ENV` in the release code is wrong, but whatever, at least our bundlers can deal with it.

Updated version:
```
if (process && typeof process.env !== "undefined" && process.env.NODE_ENV !== "production") {
```
becomes
```
if (process && typeof process.env !== "undefined" && "production" !== "production") {
```
but bundlers have nothing to do with the rest of the condition which is *wrong* because you can not check if variable exists just by doing `if (myVar) { ... }`. It throws.

As browsers don't have `process` defined in a global scope it will fail for everyone.
And there is no fix, but to revert or to follow my first comment (which is better)

**Comment by dmwmt** [comment]
And you still have another condition here:
https://github.com/mobxjs/mobx-react/blob/9850df8fd17d3f4ee7359cc3f601d14930629923/src/observerClass.js#L40
So the "fix" was not even consistent.

**Comment by dmwmt** [comment]
And when I am saying `it can not be used in a browser` it doesn't mean that I inject the script directly via the `<script>` tag.
It means, that every web application, which uses `mobx-react@6.1.5` will break, even after bundling.

**Comment by mweststrate** [comment]
@dmwmt is correct, #812 shouldn't have been merged in its current form (sorry, missed that). as the process check should indeed be preceded with `typeof process !== "undefined"`. 

Afaik that change shouldn't break dead code elimination though as `if(someCond && someOtherCond && false) {... }` should still erase the entire if block. (it might need some annotation that the conditions are side effect free, but afaik in this case it should work out of the box). In other words, I think the full check should be the following to both support substition and raw inclusion:

```javascript
if (typeof process !== "undefined" && process && process.env !== undefined && process.env.NODE_ENV !== "production") {
```

**Comment by mweststrate** [comment]
@FredyC I think the TS one will take a while, as I think it needs some extensive review, so I think its ok to accept a fix for this one first.

**Comment by FredyC** [comment]
Nah, I want to merge TS conversion today or tomorrow. As long as it passes build and tests it's alright and it can be improved over time.

I want to try out the TSDX after that which will make a correct replacement of `process.env.NODE_ENV` so there is no need for such extensive checks there. The microbundle should have that feature too, but I tried it out and doesn't seem to work.

**Comment by mweststrate** [comment]
I tried TSDX two weeks ago on immer, and it died immediately with some
obscure babel error, so I hope you'll have more luck :-P

**Comment by FredyC** [comment]
Well, I just tried TSDX against that branch and works nicely. Here is the result, no mention of `process` except in `esm` bundle which is supposed to be bundled anyway.

[dist.zip](https://github.com/mobxjs/mobx-react/files/4105266/dist.zip)

**Comment by dmwmt** [comment]
Why not rollup?

**Comment by FredyC** [comment]
Um, TSDX is Rollup, it just abstracts and makes it zero-config approach.

**Comment by FredyC** [comment]
Published 6.1.6 that should fix this issue.

**Comment by agilgur5** [comment]
Happened to check "Used By" and saw that `mobx-react` started using TSDX recently -- very cool! 😄 

@mweststrate I'm a maintainer now at TSDX and have patched or added features to most of its codebase, so can definitely help you get Immer on TSDX and debug any issues if you're still considering it! I didn't see any references to TSDX in Immer's issues or PRs so wasn't really sure what you tried (and you didn't file any issue at TSDX either! 😛 )

**Comment by FredyC** [comment]
@agilgur5 Using TSDX was my choice, Michel was kinda skeptical ;) I would love to use over at `mobx` as well, but it's not a good fit because it's sort of hand made monorepo and I don't think TSDX can handle that well.

As for Immer, I don't maintain that, but in opinion, there is no reason to break something that works, right? :)

**Comment by mweststrate** [comment]
For Immer it should be fine I'd expect, but I tried it twice and in both
cases I got very weird babel exceptions, so didn't try any further. If it
aint broken... ;-). But if you think it'd be cool if Immer uses TSDX feel
free to PR :). MobX will probably be trickier indeed.

**Comment by agilgur5** [comment]
So I just released TSDX v0.13.0 and wanted to come back to this now that that's finally out. TSDX went from 3 somewhat active maintainers to 1 (me) and I just got committing privileges last week and publishing 2 days ago -- wanted to make sure that was settled out before migrating any libraries myself.

But then I saw https://github.com/immerjs/immer/pull/536, so seems like the Immer migration was done shortly after this! Sorry I couldn't be of more help! Let me know in the future if you need any support on it!

<hr>

One major thing I noticed between the `bili` config and the TSDX config was the lack of the [`bundleNodeModules`](https://bili.egoist.sh/api/interfaces/config.html#bundlenodemodules) option. TSDX currently treats all absolute imports as externals, including for UMD (https://github.com/jaredpalmer/tsdx/issues/179), unless configured differently. That doesn't seem to impact `immer` though, as it has 0 dependencies anyway.
